torch/utils/data/datapipes/_hook_iterator.py

```diff
@@ -99,27 +106,30 @@ def _set_datapipe_valid_iterator_id(datapipe):
     return datapipe._valid_iterator_id
 
 
-def hook_iterator(namespace, profile_name):
+def hook_iterator(namespace):
     r"""
     Hook that is applied to all `__iter__` of metaclass `_DataPipeMeta`. This is done for the purpose of
     profiling and checking if an iterator is still valid.
     """
-    def profiler_record_fn_context():
-        return torch.autograd.profiler.record_function(profile_name)
+
+    def profiler_record_fn_context(datapipe):
+        if not hasattr(datapipe, "_profile_name"):
+            datapipe._profile_name = _generate_iterdatapipe_msg(datapipe, simplify_dp_name=True)
+        return torch.autograd.profiler.record_function(datapipe._profile_name)
```

Comment:
`_profile_name` is generated once for each instance of `IterDataPipe` the first time `__iter__` is called on the instance.

It is *not* shared among instances of the same class. See PR description for an example.